### PR TITLE
Add integration test test_check_downloaded_output

### PR DESCRIPTION
**Pull request by ejegrova**
Implement the integration test for checking the bundle download.

### CODE REVIEW

File: tests/integration/test_check_downloaded_output.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+import subprocess
+from os import path
+import tarfile
+from pathlib import Path
+
+import utils
+
+
+def test_check_downloaded_output(test_env):
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:11]: Would it be a good idea to reuse the completed request from `test_valid_data_in_request` to avoid needing to create a new request for each new test? I think this pattern would speed up the tests significantly over time. You could even use a pytest fixture that creates the request and then waits for it to complete. Once complete, it returns the request response. Then subsequent tests that use that fixture would just get the request response immediately.

I'll defer to your testing expertise on this though.

File: tests/integration/test_check_downloaded_output.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    assert response.status == 201
+    response = client.wait_for_complete_request(response)
+    assert response.data["state"] == "complete"
+
+    resp = client.download_bundle(response.id)
+    assert resp.status == 200
+
+    file_name = f"download_{str(response.id)}"
+    file_name_tar = f"{file_name}.tar.gz"
+    with open(file_name_tar, "wb") as file:
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:45]: Could you please download this to a temporary directory that gets automatically cleaned up after the test? Feel free to use the `tmpdir` pytest fixture to do this.

File: tests/integration/utils.py [hunk 3]
```diff
@@ -53,3 +54,56 @@ def create_new_request(self, payload):
...
+    def download_bundle(self, request_id):
+        """
+        Download a bundle archive.
+
+        :param int request_id:  ID of the request in Cachito
+        :return: Object that contains response from the Cachito API
+        :rtype: Response
+        :raises requests.exceptions.HTTPError: if the request to the Cachito API fails
+        """
+        resp = requests.get(f"{self._cachito_api_url}/requests/{request_id}/download")
```
Comment (mprahl) [inline tests/integration/utils.py:67]: These bundles can be quite large and this approach loads the bundle in memory. Could you please consider using the streaming approach as described here?
https://requests.readthedocs.io/en/latest/user/advanced/#body-content-workflow
https://stackoverflow.com/a/16696317/9706416

File: tests/integration/test_check_downloaded_output.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        file.write(resp.data)
+
+    process = subprocess.run(["gzip", "-t", file_name_tar])
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:48]: Good idea!

I'm not sure if this does the same thing, but Python has an [is_tarfile](https://docs.python.org/3/library/tarfile.html#tarfile.is_tarfile) function. If it does the same thing, it might be nice to use that instead of calling out to a command that may or may not be present on the system.

File: tests/integration/utils.py [hunk 5]
```diff
@@ -53,3 +54,56 @@ def create_new_request(self, payload):
...
+            # 300 seconds
+            if time.time() - start_time >= 300:
+                raise TimeoutError("The Cachito request did not complete within 5 minutes")
+            time.sleep(5)
+            response = self.fetch_request(response.id)
+
+        return response
+
+
+def escape_path_go(dependency):
```
Comment (mprahl) [inline tests/integration/utils.py:91]: Could you please explain why this is needed? This looks very interesting.

File: tests/integration/test_check_downloaded_output.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    names = [i["name"] for i in response.data["dependencies"]]
+    for dependency in names:
+        package_name = utils.escape_path_go(dependency)
+        dependency_path = path.join(file_name, dependencies_path, package_name)
+        assert path.exists(dependency_path)
+
+    with open(f"{file_name}/app/go.mod", "r") as file:
+        module_name = list()
+        for line in file:
+            if "module" in line:
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:64]: How about the following to be a little more precise?
```suggestion
            if line.startswith("module "):
```

I'm just concerned that some other line would have the text `module` in it such as a comment.

File: tests/integration/test_check_downloaded_output.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        tar.extractall(file_name)
+
+    dependencies_path = "deps/gomod/pkg/mod/cache/download/"
+    names = [i["name"] for i in response.data["dependencies"]]
+    for dependency in names:
+        package_name = utils.escape_path_go(dependency)
+        dependency_path = path.join(file_name, dependencies_path, package_name)
+        assert path.exists(dependency_path)
+
+    with open(f"{file_name}/app/go.mod", "r") as file:
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:61]: Optional: It might be good to assert that the path exists before trying to open it.

File: tests/integration/test_check_downloaded_output.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+                module_name.append(line.split()[-1])
+                break
+        assert module_name == [i["name"] for i in response.data["packages"]]
+
+    list_go_files = list()
+    for app_path in Path(f"./{file_name}/app/").rglob("*.go"):
+        list_go_files.append(app_path)
+    assert len(list_go_files) > 0
+
+    with tarfile.open(f"{file_name_tar}", mode="r:gz") as tar:
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:74]:
```suggestion
    with tarfile.open(file_name_tar, mode="r:gz") as tar:
```

File: tests/integration/test_check_downloaded_output.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        members = tar.getmembers()
+        path_names = set()
+        for dependency in members:
+            assert dependency.name not in path_names
+            path_names.add(dependency.name)
+
+    subprocess.run(["rm", "-rf", file_name])
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:81]: As mentioned above, it'd be better to use the `tmpdir` pytest fixture since if the test fails, these would never get called.

File: tests/integration/test_check_downloaded_output.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+import utils
+
+
+def test_check_downloaded_output(test_env):
+    """
+    Check that the bundle has all the necessities.
+
+    Process:
+    * Send new request to cachito api
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:16]:
```suggestion
    * Send new request to Cachito API
```

File: tests/integration/test_check_downloaded_output.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    * Send request to download appropriate bundle from cachito
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:17]:
```suggestion
    * Send request to download appropriate bundle from Cachito
```

File: tests/integration/test_check_downloaded_output.py (hunk 1 above)
Comment (ejegrova) [reply tests/integration/test_check_downloaded_output.py]: I didn't realise we can do it like this. Some of the scenarios will still need to create the request on their own because of the different payload (e.g. dependency replacement) but the fixture will be used for the majority of tests.

File: tests/integration/test_check_downloaded_output.py (hunk 4 above)
Comment (ejegrova) [reply tests/integration/test_check_downloaded_output.py]: It seems that it's doing the same thing. Changed it to use is_tarfile function.

File: tests/integration/test_check_downloaded_output.py (hunk 6 above)
Comment (ejegrova) [reply tests/integration/test_check_downloaded_output.py]: that's better :)

File: tests/integration/utils.py (hunk 5 above)
Comment (ejegrova) [reply tests/integration/utils.py:96]: I noticed that in the bundle the paths for dependencies changes when there is an uppercase letter. Escaping of paths is described in https://github.com/golang/mod/blob/2addee1ccfb22349ab47953a3046338e461eb4d1/module/module.go#L46
I hope it's relevant for all go modules.

**Comment by ejegrova** [comment]
@mprahl  please review that

File: tests/integration/test_check_downloaded_output.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    * Check that state is "complete"
+    * Check the downloaded data are in gzip format and valid
+    * Check that dir deps/gomod/… contains cached dependencies
+    * Check that dir app/ contains application source code
+    * Check that the same full path filename is not duplicated
+    """
+    response = create_new_request
+    assert response.status == 201
+    client = utils.Client(test_env["api_url"], test_env["api_auth_type"])
+    response = client.wait_for_complete_request(response)
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:29]: Optional: It might be nice to include this in the fixture so that all tests that use the fixture don't have to duplicate this.

In essence, the fixture would guarantee that it would supply a Cachito request that is either complete or failed.

File: tests/integration/test_check_downloaded_output.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    dependencies_path = "deps/gomod/pkg/mod/cache/download/"
+    names = [i["name"] for i in response.data["dependencies"]]
+    for dependency in names:
+        package_name = utils.escape_path_go(dependency)
+        dependency_path = path.join(file_name, dependencies_path, package_name)
+        assert path.exists(dependency_path)
+
+    assert path.exists(f"{file_name}/app/go.mod")
+    with open(f"{file_name}/app/go.mod", "r") as file:
+        module_name = list()
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:51]: When you create variables that are lists, could you make the names plural (e.g. `module_names`)? It makes it easier to read.

Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:51]: Also, please initialize lists with `module_names = []` and not `module_names = list()`.

File: tests/integration/test_check_downloaded_output.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        for line in file:
+            if line.startswith("module "):
+                module_name.append(line.split()[-1])
+                break
+        assert module_name == [i["name"] for i in response.data["packages"]]
+
+    list_go_files = list()
```
Comment (mprahl) [inline tests/integration/test_check_downloaded_output.py:58]:
```suggestion
    list_go_files = []
```

File: tests/integration/utils.py (hunk 5 above)
Comment (mprahl) [reply tests/integration/utils.py:97]: Could you add this link to the docstring? It was very helpful.

File: tests/integration/utils.py [hunk 15]
```diff
@@ -53,3 +54,62 @@ def create_new_request(self, payload):
...
+    Escape uppercase characters in names of Golang packages.
+
+    Replacing every uppercase letter with an exclamation mark followed by the lowercase letter.
+
+    :param str dependency: Name of the dependency
+    :return: Escaped dependency name
+    :rtype: str
+    """
+    if not dependency.islower():
+        package_name = str()
```
Comment (mprahl) [inline tests/integration/utils.py:108]:
```suggestion
        package_name = ''
```

**Comment by ejegrova** [comment]
@mprahl please check again :)

File: tests/integration/conftest.py [hunk 16]
```diff
@@ -18,3 +20,25 @@ def test_env():
...
+    response = client.create_new_request(
+        payload={
+            "repo": test_env["package"]["repo"],
+            "ref": test_env["package"]["ref"],
+            "pkg_managers": test_env["package"]["pkg_managers"],
+        },
+    )
+    response = client.wait_for_complete_request(response)
+
+    return response
```
Comment (mprahl) [inline tests/integration/conftest.py:44]: Optional: You could return a tuple where the first element is the initial response and second element is completed response. This way, you could use this fixture in `test_creating_new_request`.

File: tests/integration/conftest.py [hunk 17]
```diff
@@ -18,3 +20,25 @@ def test_env():
     with open(config_file) as f:
         env = yaml.safe_load(f)
     return env
+
+
+@pytest.fixture(scope="session")
+def create_new_request(test_env):
```
Comment (mprahl) [inline tests/integration/conftest.py:26]: Could you rename this to something that describes what is being returned? For example you could name it `default_request` or something like that.

I'm sorry I missed this in my previous review.

**Comment by mprahl** [comment]
Just a couple of comments. This is looking great!

File: tests/integration/conftest.py [hunk 18]
```diff
@@ -18,3 +20,25 @@ def test_env():
...
+    initial_response = client.create_new_request(
+        payload={
+            "repo": test_env["package"]["repo"],
+            "ref": test_env["package"]["ref"],
+            "pkg_managers": test_env["package"]["pkg_managers"],
+        },
+    )
+    completed_response = client.wait_for_complete_request(initial_response)
+
+    return initial_response, completed_response
```
Comment (mprahl) [inline tests/integration/conftest.py:44]: Sorry again for not mentioning this earlier, but perhaps a named tuple would make this easier to use:
```python
from collections import namedtuple
DefaultRequest = namedtuple('DefaultRequest', 'initial_response complete_response')
```

Then in your tests, you can use it like `default_request.initial_response` and `default_request.complete_response`

File: tests/integration/conftest.py [hunk 19]
```diff
@@ -18,3 +20,25 @@ def test_env():
...
     return env
+
+
+@pytest.fixture(scope="session")
+def default_request(test_env):
+    """
+    Create a new request in Cachito.
+
+    :param test_env: Test environment configuration
+    :return: Object that contains initial and completed response from the Cachito API
```
Comment (mprahl) [inline tests/integration/conftest.py:31]:
```suggestion
    :return: a tuple that contains initial and completed response from the Cachito API
```

**Comment by mprahl** [comment]
@ejegrova after these two comments are addressed, this is good to merge. Great work!

**Comment by ejegrova** [comment]
Done the changes, @mprahl please check
